// @sanity/form-builder/lib/inputs/PortableText/__workshop__/default/Story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Story;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _uiWorkshop = require("@sanity/ui-workshop");

var _TestInput = require("../TestInput");

var _schema = require("./schema");

var _values = require("./values");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Story() {
  var readOnly = (0, _uiWorkshop.useBoolean)('Read only', false);
  var selectedValue = (0, _uiWorkshop.useSelect)('Values', _values.valueOptions) || 'empty';
  var value = _values.values[selectedValue];
  return /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_TestInput.TestInput, {
    readOnly: readOnly,
    value: value,
    type: _schema.portableTextType,
    schema: _schema.schema
  })))));
}
