// @sanity/desk-tool/lib/panes/unknown/UnknownPaneType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownPane = UnknownPane;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _pane = require("../../components/pane");

var _typePredicates = require("../../utils/typePredicates");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @internal
 */
function UnknownPane(props) {
  var index = props.index,
      isSelected = props.isSelected,
      pane = props.pane;
  var type = (0, _typePredicates.isRecord)(pane) && pane.type || null;
  return /*#__PURE__*/_react.default.createElement(_pane.Pane, {
    "data-index": index,
    selected: isSelected
  }, /*#__PURE__*/_react.default.createElement(_pane.PaneHeader, {
    title: "Unknown pane type"
  }), /*#__PURE__*/_react.default.createElement(_pane.PaneContent, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, typeof type === 'string' ? /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Structure item of type ", /*#__PURE__*/_react.default.createElement("code", null, type), " is not a known entity.") : /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Structure item is missing required ", /*#__PURE__*/_react.default.createElement("code", null, "type"), " property."))));
}
