// @sanity/form-builder/lib/inputs/PortableText/__workshop__/custom/schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = exports.portableTextType = void 0;

var _schema = _interopRequireDefault(require("@sanity/schema"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var imageType = {
  type: 'image',
  name: 'blockImage'
};
var someObject = {
  type: 'object',
  name: 'someObject',
  fields: [{
    type: 'string',
    name: 'color'
  }]
};
var blockType = {
  type: 'block',
  name: 'myBlockType',
  styles: [{
    title: 'Normal',
    value: 'normal'
  }, {
    title: 'H1',
    value: 'h1'
  }, {
    title: 'H2',
    value: 'h2'
  }, {
    title: 'H3',
    value: 'h3'
  }, {
    title: 'H4',
    value: 'h4'
  }, {
    title: 'H5',
    value: 'h5'
  }, {
    title: 'H6',
    value: 'h6'
  }, {
    title: 'Quote',
    value: 'blockquote'
  }],
  of: [someObject, imageType]
};
var ptType = {
  type: 'array',
  name: 'body',
  of: [blockType, imageType]
};

var schema = _schema.default.compile({
  name: 'default',
  types: [imageType, someObject, ptType]
});

exports.schema = schema;
var portableTextType = schema.get('body');
exports.portableTextType = portableTextType;
