// @sanity/desk-tool/lib/components/pane/PaneHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneHeader = void 0;

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _usePane2 = require("./usePane");

var _PaneHeader = require("./PaneHeader.styles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * @beta This API will change. DO NOT USE IN PRODUCTION.
 */
var PaneHeader = /*#__PURE__*/(0, _react.forwardRef)(function PaneHeader(props, ref) {
  var actions = props.actions,
      backButton = props.backButton,
      loading = props.loading,
      subActions = props.subActions,
      tabs = props.tabs,
      title = props.title;

  var _usePane = (0, _usePane2.usePane)(),
      collapse = _usePane.collapse,
      collapsed = _usePane.collapsed,
      expand = _usePane.expand,
      paneElement = _usePane.rootElement;

  var paneRect = (0, _ui.useElementRect)(paneElement || null);
  var collapsedRef = (0, _react.useRef)(collapsed);

  var _useState = (0, _react.useState)(!collapsed),
      _useState2 = _slicedToArray(_useState, 2),
      actionsVisible = _useState2[0],
      setActionsVisible = _useState2[1];

  (0, _react.useEffect)(() => {
    collapsedRef.current = collapsed;
    if (!collapsed) return;
    setActionsVisible(false);
  }, [collapsed]);
  var layoutStyle = (0, _react.useMemo)(() => {
    return {
      width: collapsed ? (paneRect === null || paneRect === void 0 ? void 0 : paneRect.height) || window.innerHeight : undefined
    };
  }, [collapsed, paneRect]);
  var handleTransitionEnd = (0, _react.useCallback)(() => {
    if (collapsedRef.current) return;
    setActionsVisible(true);
  }, []);
  var handleTitleClick = (0, _react.useCallback)(() => {
    if (collapsed) return;
    collapse();
  }, [collapse, collapsed]);
  var handleLayoutClick = (0, _react.useCallback)(() => {
    if (!collapsed) return;
    expand();
  }, [collapsed, expand]);
  var titleBox = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleBox, {
    flex: 1,
    onClick: handleTitleClick,
    paddingY: 3,
    paddingLeft: backButton ? 1 : 3
  }, loading && /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleTextSkeleton, {
    animated: true,
    radius: 1
  }), !loading && /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleText, {
    tabIndex: 0,
    textOverflow: "ellipsis",
    weight: "semibold"
  }, title)), [backButton, handleTitleClick, loading, title]);
  var row2 = (0, _react.useMemo)(() => (tabs || subActions) && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    hidden: collapsed,
    paddingTop: 0,
    paddingRight: 2,
    paddingBottom: 2,
    paddingLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.TabsBox, {
    flex: 1,
    marginRight: subActions ? 3 : 0
  }, /*#__PURE__*/_react.default.createElement("div", null, tabs)), subActions && /*#__PURE__*/_react.default.createElement(_ui.Box, null, subActions)), [collapsed, subActions, tabs]);
  var layout = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_PaneHeader.Layout, {
    onTransitionEnd: handleTransitionEnd,
    onClick: handleLayoutClick,
    padding: 2,
    paddingBottom: tabs || subActions ? 1 : 2,
    sizing: "border",
    style: layoutStyle
  }, backButton, titleBox, actions && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    hidden: !actionsVisible,
    paddingLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "paneHeader"
  }, actions))), [actions, actionsVisible, backButton, handleLayoutClick, handleTransitionEnd, layoutStyle, subActions, tabs, titleBox]);
  return (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, {
    zOffset: 100
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.Root, {
    "data-collapsed": collapsed ? '' : undefined,
    "data-testid": "pane-header",
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "paneHeader"
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-collapsed": collapsed ? '' : undefined,
    tone: "inherit"
  }, layout, row2)))), [collapsed, layout, ref, row2]);
});
exports.PaneHeader = PaneHeader;
