// @sanity/desk-tool/lib/panes/documentList/CreateMenuButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateMenuButton = CreateMenuButton;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _IntentMenuItem = require("../../components/IntentMenuItem");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var POPOVER_PROPS = {
  constrainSize: true,
  placement: 'bottom',
  portal: true
};

function CreateMenuButton(props) {
  var items = props.items;
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.ComposeIcon,
      mode: "bleed",
      padding: 3
    }),
    id: "create-menu",
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 3,
      paddingTop: 3,
      paddingBottom: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
      muted: true
    }, "Create")), items.map((createItem, createItemIndex) => /*#__PURE__*/_react.default.createElement(_IntentMenuItem.IntentMenuItem, {
      icon: createItem.icon,
      intent: createItem.intent,
      key: createItemIndex,
      text: createItem.title
    }))),
    popover: POPOVER_PROPS
  });
}
