// @sanity/desk-tool/lib/getIntentState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getIntentState = getIntentState;
exports.setActivePanes = setActivePanes;

var _uuid = require("@sanity/uuid");

var _constants = require("./constants");

var state = {
  activePanes: []
};

function setActivePanes(panes) {
  state.activePanes = panes;
}
/**
 * @internal
 */


function getIntentState(intentName, params, currentState, payload) {
  var panes = (currentState === null || currentState === void 0 ? void 0 : currentState.panes) || [];
  var activePanes = state.activePanes || [];
  var editDocumentId = params.id || (0, _uuid.uuid)();
  var isTemplate = intentName === 'create' && params.template; // Loop through open panes and see if any of them can handle the intent

  for (var i = activePanes.length - 1; i >= 0; i--) {
    var _pane$canHandleIntent;

    var pane = activePanes[i];

    if (typeof pane === 'object' && (_pane$canHandleIntent = pane.canHandleIntent) !== null && _pane$canHandleIntent !== void 0 && _pane$canHandleIntent.call(pane, intentName, params, {
      // @ts-expect-error: this was incorrectly typed as `never`
      // because TS has trouble with inferred intersections
      pane,
      index: i
    })) {
      var paneParams = isTemplate ? {
        template: params.template
      } : _constants.EMPTY_PARAMS;
      return {
        panes: panes.slice(0, i).concat([[{
          id: editDocumentId,
          params: paneParams,
          payload
        }]])
      };
    }
  }

  return {
    intent: intentName,
    params,
    payload
  };
}
