// @sanity/desk-tool/lib/panes/document/documentPanel/header/ValidationMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidationMenu = ValidationMenu;

var _autoId = require("@reach/auto-id");

var _components = require("@sanity/base/components");

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _useDocumentPane2 = require("../../useDocumentPane");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ValidationMenu(props) {
  var boundaryElement = props.boundaryElement,
      isOpen = props.isOpen,
      setOpen = props.setOpen;

  var _useDocumentPane = (0, _useDocumentPane2.useDocumentPane)(),
      documentSchema = _useDocumentPane.documentSchema,
      handleFocus = _useDocumentPane.handleFocus,
      markers = _useDocumentPane.markers;

  var validationMarkers = (0, _react.useMemo)(() => markers.filter(marker => marker.type === 'validation'), [markers]);
  var validationErrorMarkers = (0, _react.useMemo)(() => validationMarkers.filter(marker => marker.level === 'error'), [validationMarkers]);
  var validationWarningwarnings = (0, _react.useMemo)(() => validationMarkers.filter(marker => marker.level === 'warning'), [validationMarkers]);
  var id = (0, _autoId.useId)();
  var handleClose = (0, _react.useCallback)(() => setOpen(false), [setOpen]);

  if (validationErrorMarkers.length === 0 && validationWarningwarnings.length === 0) {
    return null;
  }

  var popoverContent = /*#__PURE__*/_react.default.createElement(_components.ValidationList, {
    documentType: documentSchema,
    markers: validationMarkers,
    onClose: handleClose,
    onFocus: handleFocus
  });

  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    id: id || '',
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.ErrorOutlineIcon,
      title: "Show validation issues",
      mode: "bleed",
      tone: "critical"
    }),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, {
      open: isOpen
    }, popoverContent),
    popover: {
      portal: true,
      boundaryElement: boundaryElement,
      constrainSize: true,
      preventOverflow: true,
      width: 0
    },
    placement: "bottom-end"
  });
}
