// @sanity/desk-tool/lib/components/pane/__workshop__/resize.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ResizeStory;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _ = require("..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ResizeStory() {
  return /*#__PURE__*/_react.default.createElement(_.PaneLayout, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(ListPane, null), /*#__PURE__*/_react.default.createElement(ListPane, null), /*#__PURE__*/_react.default.createElement(DocumentPane, null));
}

function ListPane() {
  return /*#__PURE__*/_react.default.createElement(_.Pane, {
    currentMaxWidth: 350,
    flex: 1,
    minWidth: 320 / 2,
    maxWidth: 500
  }, /*#__PURE__*/_react.default.createElement(_.PaneHeader, {
    title: "List"
  }));
}

function DocumentPane() {
  return /*#__PURE__*/_react.default.createElement(_.Pane, {
    flex: 2,
    minWidth: 320 / 2
  }, /*#__PURE__*/_react.default.createElement(_.PaneHeader, {
    title: "Document"
  }), /*#__PURE__*/_react.default.createElement(_.PaneContent, null), /*#__PURE__*/_react.default.createElement(_.PaneFooter, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "Footer")));
}
