// @sanity/base/lib/components/__workshop__/CollapseMenuStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CollapseMenuStory;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _ = require("..");

var _FormatUnderlined = _interopRequireDefault(require("../icons/FormatUnderlined"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CollapseMenuStory() {
  return /*#__PURE__*/_react.default.createElement(_ui.Container, {
    padding: 4,
    width: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    shadow: 1,
    padding: 1,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_.CollapseMenu, null, /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _icons.BoldIcon,
    text: "Bold",
    buttonProps: {
      mode: 'bleed'
    },
    tooltipProps: {
      placement: 'right'
    }
  }), /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _icons.ItalicIcon,
    text: "Italic",
    buttonProps: {
      mode: 'bleed'
    }
  }), /*#__PURE__*/_react.default.createElement(_.CollapseMenuDivider, null), /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _icons.CodeIcon,
    text: "Code",
    buttonProps: {
      mode: 'bleed'
    }
  }), /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _FormatUnderlined.default,
    text: "Underline",
    buttonProps: {
      mode: 'bleed'
    }
  }), /*#__PURE__*/_react.default.createElement(_.CollapseMenuDivider, null), /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _icons.UlistIcon,
    text: "Ulist",
    buttonProps: {
      mode: 'bleed'
    }
  }), /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    icon: _icons.OlistIcon,
    text: "Olist",
    buttonProps: {
      mode: 'bleed'
    },
    selected: true
  }))));
}
