// @sanity/desk-tool/lib/contexts/paneRouter/PaneRouterProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneRouterProvider = PaneRouterProvider;

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _pick2 = _interopRequireDefault(require("lodash/pick"));

var _router = require("@sanity/base/router");

var _react = _interopRequireWildcard(require("react"));

var _constants = require("./constants");

var _ChildLink = require("./ChildLink");

var _PaneRouterContext = require("./PaneRouterContext");

var _ParameterizedLink = require("./ParameterizedLink");

var _excluded = ["view"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var DEFAULT_SET_PARAMS_OPTIONS = {
  recurseIfInherited: false
};
/**
 * @internal
 */

function PaneRouterProvider(props) {
  var children = props.children,
      flatIndex = props.flatIndex,
      index = props.index,
      paramsProp = props.params,
      payloadProp = props.payload,
      siblingIndex = props.siblingIndex;

  var _useRouter = (0, _router.useRouter)(),
      navigate = _useRouter.navigate,
      navigateIntent = _useRouter.navigateIntent;

  var routerState = (0, _router.useRouterState)();
  var routerPaneGroups = (0, _react.useMemo)(() => (routerState === null || routerState === void 0 ? void 0 : routerState.panes) || [], [routerState === null || routerState === void 0 ? void 0 : routerState.panes]);
  var groupIndex = index - 1;

  var _useState = (0, _react.useState)(paramsProp),
      _useState2 = _slicedToArray(_useState, 2),
      params = _useState2[0],
      _setParams = _useState2[1];

  var paramsRef = (0, _react.useRef)(paramsProp);

  var _useState3 = (0, _react.useState)(payloadProp),
      _useState4 = _slicedToArray(_useState3, 2),
      payload = _useState4[0],
      _setPayload = _useState4[1];

  var payloadRef = (0, _react.useRef)(payloadProp); // Update params state

  (0, _react.useEffect)(() => {
    if (!(0, _isEqual2.default)(paramsRef.current, paramsProp)) {
      paramsRef.current = paramsProp;

      _setParams(paramsProp);
    }
  }, [paramsProp]); // Update payload state

  (0, _react.useEffect)(() => {
    if (!(0, _isEqual2.default)(payloadRef.current, payloadProp)) {
      payloadRef.current = payloadProp;

      _setPayload(payloadProp);
    }
  }, [payloadProp]);
  var modifyCurrentGroup = (0, _react.useCallback)(modifier => {
    var newPanes = routerPaneGroups.slice(0);
    var currentGroup = routerPaneGroups[groupIndex] ? routerPaneGroups[groupIndex].slice(0) : [];
    newPanes.splice(groupIndex, 1, modifier(currentGroup, currentGroup[siblingIndex]));

    var newRouterState = _objectSpread(_objectSpread({}, routerState || {}), {}, {
      panes: newPanes
    });

    setTimeout(() => navigate(newRouterState), 0);
    return newRouterState;
  }, [groupIndex, navigate, routerPaneGroups, routerState, siblingIndex]);
  var setPayload = (0, _react.useCallback)(nextPayload => {
    var currPayload = payloadRef.current;

    if (!(0, _isEqual2.default)(currPayload, nextPayload)) {
      _setPayload(nextPayload);

      payloadRef.current = nextPayload;
    }

    modifyCurrentGroup((siblings, item) => {
      var newGroup = siblings.slice();
      newGroup[siblingIndex] = _objectSpread(_objectSpread({}, item), {}, {
        payload: nextPayload
      });
      return newGroup;
    });
  }, [modifyCurrentGroup, siblingIndex]);
  var setParams = (0, _react.useCallback)(function (nextParams) {
    var setOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var currParams = paramsRef.current;

    if (!(0, _isEqual2.default)(currParams, nextParams)) {
      _setParams(nextParams);

      paramsRef.current = nextParams;
    }

    var _DEFAULT_SET_PARAMS_O = _objectSpread(_objectSpread({}, DEFAULT_SET_PARAMS_OPTIONS), setOptions),
        recurseIfInherited = _DEFAULT_SET_PARAMS_O.recurseIfInherited;

    modifyCurrentGroup((siblings, item) => {
      var isGroupRoot = siblingIndex === 0;
      var isDuplicate = !isGroupRoot && item.id === siblings[0].id;
      var newGroup = siblings.slice();

      if (!isDuplicate) {
        newGroup[siblingIndex] = _objectSpread(_objectSpread({}, item), {}, {
          params: nextParams
        });
        return newGroup;
      }

      var rootParams = siblings[0].params;

      if (recurseIfInherited) {
        var newParamKeys = Object.keys(nextParams);
        var inheritedKeys = Object.keys(paramsProp).filter(key => rootParams && rootParams[key] === paramsProp[key]);
        var removedInheritedKeys = inheritedKeys.filter(key => !nextParams[key]);
        var remainingInheritedKeys = newParamKeys.filter(key => inheritedKeys.includes(key));
        var exclusiveKeys = newParamKeys.filter(key => !inheritedKeys.includes(key));
        var exclusive = (0, _pick2.default)(nextParams, exclusiveKeys);

        var inherited = _objectSpread(_objectSpread({}, (0, _omit2.default)(rootParams, removedInheritedKeys)), (0, _pick2.default)(nextParams, remainingInheritedKeys));

        newGroup[0] = _objectSpread(_objectSpread({}, item), {}, {
          params: inherited
        });
        newGroup[siblingIndex] = _objectSpread(_objectSpread({}, item), {}, {
          params: exclusive
        });
      } else {
        // If it's a duplicate of the group root, we should only set the parameters
        // that differ from the group root.
        var newParams = Object.keys(nextParams).reduce((siblingParams, key) => {
          if (_constants.exclusiveParams.includes(key) || rootParams && nextParams[key] !== rootParams[key]) {
            siblingParams[key] = nextParams[key];
          }

          return siblingParams;
        }, {});
        newGroup[siblingIndex] = _objectSpread(_objectSpread({}, item), {}, {
          params: newParams
        });
      }

      return newGroup;
    });
  }, [modifyCurrentGroup, paramsProp, siblingIndex]);
  var ctx = (0, _react.useMemo)(() => ({
    // Zero-based index (position) of pane, visually
    index: flatIndex,
    // Zero-based index of pane group (within URL structure)
    groupIndex,
    // Zero-based index of pane within sibling group
    siblingIndex,
    // Payload of the current pane
    payload,
    // Params of the current pane
    params,
    // Whether or not the pane has any siblings (within the same group)
    hasGroupSiblings: routerPaneGroups[groupIndex] ? routerPaneGroups[groupIndex].length > 1 : false,
    // The length of the current group
    groupLength: routerPaneGroups[groupIndex] ? routerPaneGroups[groupIndex].length : 0,
    // Current router state for the "panes" property
    routerPanesState: routerPaneGroups,
    // Curried StateLink that passes the correct state automatically
    ChildLink: _ChildLink.ChildLink,
    // Curried StateLink that passed the correct state, but merges params/payload
    ParameterizedLink: _ParameterizedLink.ParameterizedLink,
    // Replaces the current pane with a new one
    replaceCurrent: function replaceCurrent() {
      var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      modifyCurrentGroup(() => [{
        id: opts.id || '',
        payload: opts.payload,
        params: opts.params || {}
      }]);
    },
    // Removes the current pane from the group
    closeCurrent: () => {
      modifyCurrentGroup((siblings, item) => siblings.length > 1 ? siblings.filter(sibling => sibling !== item) : siblings);
    },
    // Duplicate the current pane, with optional overrides for payload, parameters
    duplicateCurrent: options => {
      // const {payload, params} = options || {}
      modifyCurrentGroup((siblings, item) => {
        var newGroup = siblings.slice();
        newGroup.splice(siblingIndex + 1, 0, _objectSpread(_objectSpread({}, item), {}, {
          payload: (options === null || options === void 0 ? void 0 : options.payload) || item.payload,
          params: (options === null || options === void 0 ? void 0 : options.params) || item.params
        }));
        return newGroup;
      });
    },
    // Set the view for the current pane
    setView: viewId => {
      var _paramsRef$current = paramsRef.current,
          view = _paramsRef$current.view,
          restParams = _objectWithoutProperties(_paramsRef$current, _excluded);

      return setParams(viewId ? _objectSpread(_objectSpread({}, restParams), {}, {
        view: viewId
      }) : restParams);
    },
    // Set the parameters for the current pane
    setParams,
    // Set the payload for the current pane
    setPayload,
    // Proxied navigation to a given intent. Consider just exposing `router` instead?
    navigateIntent
  }), [flatIndex, groupIndex, modifyCurrentGroup, navigateIntent, params, payload, routerPaneGroups, setParams, setPayload, siblingIndex]);
  return /*#__PURE__*/_react.default.createElement(_PaneRouterContext.PaneRouterContext.Provider, {
    value: ctx
  }, children);
}
